### PR TITLE
RTEMIS-137: Added the parent fields for Student Details Form.

### PR DESCRIPTION
**Pull request by RohanSingh-inno**
Ticket # : https://innoraft.atlassian.net/browse/RTEMIS-137

---
### What the PR does
`(Explain fixes, changes and impacts)`
- Added parent detail fields and added respective validations.

---
### Extras
https://www.awesomescreenshot.com/video/27540934?key=80719f393c356f1bcd6a4f696bd51fd1

### CODE REVIEW

**Review by jaswinsingh — COMMENTED** [review]
@RohanSingh-inno Please add screen-recording in the PR.

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_father_aadhar_number.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+dependencies:
+  config:
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_father_aadhar_number
+id: mini_node.student_details.field_father_aadhar_number
+field_name: field_father_aadhar_number
+entity_type: mini_node
+bundle: student_details
+label: "Father's Aadhar Number"
+description: ''
```
Comment (jaswinsingh) [inline modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_father_aadhar_number.yml:12]: Please add an appropriate description here.

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_father_name.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+dependencies:
+  config:
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_father_name
+id: mini_node.student_details.field_father_name
+field_name: field_father_name
+entity_type: mini_node
+bundle: student_details
+label: "Father's Name"
+description: ''
```
Comment (jaswinsingh) [inline modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_father_name.yml:12]: Please add an appropriate description here.

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_gaurdian_aadhar_number.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+dependencies:
+  config:
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_gaurdian_aadhar_number
+id: mini_node.student_details.field_gaurdian_aadhar_number
+field_name: field_gaurdian_aadhar_number
+entity_type: mini_node
+bundle: student_details
+label: "Gaurdian's Aadhar Number"
+description: ''
```
Comment (jaswinsingh) [inline modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_gaurdian_aadhar_number.yml:12]: Please add an appropriate description here.

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_guardian_name.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+dependencies:
+  config:
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_guardian_name
+id: mini_node.student_details.field_guardian_name
+field_name: field_guardian_name
+entity_type: mini_node
+bundle: student_details
+label: "Guardian's Name"
+description: ''
```
Comment (jaswinsingh) [inline modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_guardian_name.yml:12]: Please add an appropriate description here.

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_mother_aadhar_number.yml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+dependencies:
+  config:
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_mother_aadhar_number
+id: mini_node.student_details.field_mother_aadhar_number
+field_name: field_mother_aadhar_number
+entity_type: mini_node
+bundle: student_details
+label: "Mother's Aadhar Number"
+description: ''
```
Comment (jaswinsingh) [inline modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_mother_aadhar_number.yml:12]: Please add an appropriate description here.

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_mother_name.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+dependencies:
+  config:
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_mother_name
+id: mini_node.student_details.field_mother_name
+field_name: field_mother_name
+entity_type: mini_node
+bundle: student_details
+label: "Mother's Name"
+description: ''
```
Comment (jaswinsingh) [inline modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_mother_name.yml:12]: Please add an appropriate description here.

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_parent_type.yml [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_parent_type
+  module:
+    - options
+id: mini_node.student_details.field_parent_type
+field_name: field_parent_type
+entity_type: mini_node
+bundle: student_details
+label: 'Parent Type'
+description: ''
```
Comment (jaswinsingh) [inline modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_parent_type.yml:14]: Please add an appropriate description here.

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_single_parent_type.yml [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_single_parent_type
+  module:
+    - options
+id: mini_node.student_details.field_single_parent_type
+field_name: field_single_parent_type
+entity_type: mini_node
+bundle: student_details
+label: 'Single Parent Type'
+description: ''
```
Comment (jaswinsingh) [inline modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_single_parent_type.yml:14]: Please add an appropriate description here.

File: modules/rte_mis_student/rte_mis_student.module [hunk 9]
```diff
@@ -327,6 +327,35 @@ function rte_mis_student_validate_mini_node_student($form, FormStateInterface $f
...
+      }
+    }
+    elseif ($values['field_parent_type'][0]['value'] == 'guardian') {
+      $form_state->setValue('field_single_parent_type', ['0' => ['value' => NULL]]);
+      $form_state->setValue('field_father_name', ['0' => ['value' => '']]);
+      $form_state->setValue('field_father_aadhar_number', ['0' => ['value' => '']]);
+      $form_state->setValue('field_mother_name', ['0' => ['value' => '']]);
+      $form_state->setValue('field_mother_aadhar_number', ['0' => ['value' => '']]);
+    }
+  }
```
Comment (jaswinsingh) [inline modules/rte_mis_student/rte_mis_student.module:358]: Can we use `form_state->getValues` at once unset the values in an array and then assign the values using `form_state->setValues`?

Comment (RohanSingh-inno) [reply modules/rte_mis_student/rte_mis_student.module]: ok

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_single_parent_type.yml (hunk 8 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_single_parent_type.yml]: ok

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_parent_type.yml (hunk 7 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_parent_type.yml]: ok

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_mother_name.yml (hunk 6 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_mother_name.yml]: ok

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_mother_aadhar_number.yml (hunk 5 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_mother_aadhar_number.yml]: ok

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_guardian_name.yml (hunk 4 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_guardian_name.yml]: ok

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_gaurdian_aadhar_number.yml (hunk 3 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_gaurdian_aadhar_number.yml]: ok

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_father_name.yml (hunk 2 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_father_name.yml]: ok

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_father_aadhar_number.yml (hunk 1 above)
Comment (RohanSingh-inno) [reply modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_father_aadhar_number.yml]: ok

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_mother_aadhar_number.yml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+dependencies:
+  config:
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_mother_aadhar_number
+id: mini_node.student_details.field_mother_aadhar_number
+field_name: field_mother_aadhar_number
+entity_type: mini_node
+bundle: student_details
+label: "Mother's Aadhar Number"
+description: "Student's Mother's Aadhar Card Number"
```
Comment (jaswinsingh) [inline modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_mother_aadhar_number.yml:12]: Same here Description should mention to add the last 4 digit number of aadhaar 
Why is this field of type string? user can enter string values?

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_gaurdian_aadhar_number.yml [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+dependencies:
+  config:
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_gaurdian_aadhar_number
+id: mini_node.student_details.field_gaurdian_aadhar_number
+field_name: field_gaurdian_aadhar_number
+entity_type: mini_node
+bundle: student_details
+label: "Gaurdian's Aadhar Number"
+description: "Student's Gaurdian's Aadhar Card Number"
```
Comment (jaswinsingh) [inline modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_gaurdian_aadhar_number.yml:12]: Same here, Description should mention to add the last 4 digit number of aadhaar 
Why is this field of type string? user can enter string values?

File: modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_father_aadhar_number.yml [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+dependencies:
+  config:
+    - eck.eck_type.mini_node.student_details
+    - field.storage.mini_node.field_father_aadhar_number
+id: mini_node.student_details.field_father_aadhar_number
+field_name: field_father_aadhar_number
+entity_type: mini_node
+bundle: student_details
+label: "Father's Aadhar Number"
+description: "Student's Father's Aadhar Card Number"
```
[the remark this answers is not in the document]
Comment (RohanSingh-inno) [reply modules/rte_mis_student/config/install/field.field.mini_node.student_details.field_father_aadhar_number.yml:12]: I had a second thought about keeping it integer, but what if the last 4 digits are '0001' the integer field will only store 1. We can keep a constraint that the user can enter only integer values.

**Review by jaswinsingh — APPROVED** [review]

**Pull request merged**